### PR TITLE
Dashboard: Mix initials and custom gravatars in recent viewers list

### PR DESCRIPTION
**Pull request by mmandrus**
**What this PR does / why we need it**:
This PR augments the enterprise "recent viewers" feature when looking at dashboards. 

Previously, we displayed either **only** initials or **only** gravatars, depending on the disable_gravatar flag. This change allows for mixing of custom avatars (i.e. set by the user) and initials for "default" avatars (i.e. the retro-style fallback avatars). Because default avatars are random, this change allows users to more easily understand who is viewing dashboards at a glance.

Screenshot_2022-04-01 11 59 52_zEwd0j

**Which issue(s) this PR fixes**: 
grafana-enterprise/544

**Other**
- This has a corresponding PR on the grafana-enterprise side
- Thank you @pianohacker for helping decide the implementation approach

### CODE REVIEW

**Comment by mmandrus** [comment]
Has a failing front end unit test due to the backend change, update coming soon

**Review by leandro-deveikis — APPROVED** [review]
LGTM!

File: pkg/api/avatar/avatar.go [hunk 1]
```diff
@@ -214,11 +259,48 @@ var client = &http.Client{
...
+		return performGet(a.BaseUrl+hasCustomReqParams, a.Avatar, checkIsCustomHandler)
+	}
+	return err
+}
+
+type ResponseHandler func(av *Avatar, resp *http.Response) error
+
+// Verifies the Gravatar response code was 200, then stores the image byte slice
+func getGravatarHandler(av *Avatar, resp *http.Response) error {
+	if resp.StatusCode != 200 {
```
Comment (evictorero) [inline pkg/api/avatar/avatar.go:283]:
```suggestion
	if resp.StatusCode != http.StatusOK {
```

File: pkg/api/avatar/avatar.go [hunk 2]
```diff
@@ -214,11 +259,48 @@ var client = &http.Client{
...
+	av.data = &bytes.Buffer{}
+	writer := bufio.NewWriter(av.data)
+
+	_, err := io.Copy(writer, resp.Body)
+	return err
+}
+
+// Uses the d=404 fallback to see if the gravatar we got back is custom
+func checkIsCustomHandler(av *Avatar, resp *http.Response) error {
+	av.isCustom = resp.StatusCode != 404
```
Comment (evictorero) [inline pkg/api/avatar/avatar.go:297]:
```suggestion
	av.isCustom = resp.StatusCode != http.StatusNotFound
```

File: pkg/api/avatar/avatar.go [hunk 3]
```diff
@@ -227,9 +309,10 @@ func (a *thunderTask) fetch() error {
 	req.Header.Set("Accept-Language", "zh-CN,zh;q=0.8")
 	req.Header.Set("Cache-Control", "no-cache")
 	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/33.0.1750.154 Safari/537.36")
+	alog.Debug("Fetching avatar url with parameters", "url", url)
 	resp, err := client.Do(req)
 	if err != nil {
-		a.Avatar.notFound = true
+		av.HandleErrorResponse()
```
Comment (evictorero) [inline pkg/api/avatar/avatar.go:315]: If an error occurs we return an error. Is it really needed to set the `Avatar` values before returning an error?

**Review by evictorero — COMMENTED** [review]
Great work @mmandrus ! Love this feature, makes a lot of sense. I left some comments and nits.

File: pkg/api/avatar/avatar.go [hunk 4]
```diff
@@ -61,31 +70,67 @@ func (a *Avatar) Encode(wr io.Writer) error {
...
 }
 
-func (a *Avatar) Update() (err error) {
+func (a *Avatar) update(baseUrl string) (err error) {
+	baseUrl += a.hash + "?"
 	select {
 	case <-time.After(time.Second * 3):
 		err = fmt.Errorf("get gravatar image %s timeout", a.hash)
-	case err = <-thunder.GoFetch(gravatarSource+a.hash+"?"+a.reqParams, a):
+	case err = <-thunder.GoFetch(baseUrl, a):
```
[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/api/avatar/avatar.go:78]: Yes, the params get appended as needed in the thunderTask.fetch() function

File: pkg/api/avatar/avatar.go [hunk 5]
```diff
@@ -61,31 +70,67 @@ func (a *Avatar) Encode(wr io.Writer) error {
 	return err
 }
 
-func (a *Avatar) Update() (err error) {
+func (a *Avatar) update(baseUrl string) (err error) {
+	baseUrl += a.hash + "?"
```
[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/api/avatar/avatar.go:74]: okay, will assign to a new variable

File: pkg/api/avatar/avatar.go (hunk 3 above)
Comment (mmandrus) [reply pkg/api/avatar/avatar.go:315]: In this case, yes. The flags we set on the avatar tell the endpoint to return the default "not found" image. We could use default values instead, but I think setting them explicitly every time is safer.

**Review by evictorero — APPROVED** [review]
👌

File: pkg/api/dtos/models.go [hunk 6]
```diff
@@ -78,11 +78,20 @@ func GetGravatarUrl(text string) string {
...
+	if text == "" {
+		return make([]byte, 0), false
+	}
+
 	hasher := md5.New()
 	if _, err := hasher.Write([]byte(strings.ToLower(text))); err != nil {
 		mlog.Warn("Failed to hash text", "err", err)
 	}
-	return fmt.Sprintf(setting.AppSubUrl+"/avatar/%x", hasher.Sum(nil))
+	return hasher.Sum(nil), true
```
Comment (jalevin) [inline pkg/api/dtos/models.go:94]: Would it be better to return an empty byte slice rather than performing sum operation on nil?

**Pull request merged**